File: routes/biz.py
"""BIZ 业务数据查询路由 — adpilot_biz 归一化数据"""
from __future__ import annotations

import asyncio
import re
from typing import Optional

from fastapi import APIRouter, Query, HTTPException

from repositories import (
    biz_daily_report_repository,
    biz_campaign_repository,
    biz_adgroup_daily_repository,
    biz_ad_daily_repository,
    biz_adgroup_repository,
    biz_ad_repository,
)

router = APIRouter(prefix="/biz", tags=["BIZ 业务数据"])

_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _check_dates(start_date: str, end_date: str):
    if not _DATE_RE.match(start_date):
        raise HTTPException(400, f"start_date 格式错误，应为 YYYY-MM-DD，实际: {start_date}")
    if not _DATE_RE.match(end_date):
        raise HTTPException(400, f"end_date 格式错误，应为 YYYY-MM-DD，实际: {end_date}")
    if start_date > end_date:
        raise HTTPException(400, f"start_date({start_date}) 不能大于 end_date({end_date})")


@router.get("/campaigns")
async def biz_campaigns(
    platform: Optional[str] = Query(None, description="平台过滤: tiktok / meta"),
):
    """返回 BIZ 库中所有 campaign（含 status / account_id / raw_json）"""
    if platform:
        rows = await asyncio.to_thread(biz_campaign_repository.list_by_platform, platform)
    else:
        rows = await asyncio.to_thread(biz_campaign_repository.list_all)
    for row in rows:
        for k in ("created_at", "updated_at"):
            if k in row and row[k] is not None:
                row[k] = str(row[k])
    return {"code": 0, "message": "ok", "data": rows}


@router.post("/update-status")
async def biz_update_status(
    platform: str = Query(..., description="tiktok / meta"),
    entity_type: str = Query(..., description="campaign / adgroup / ad"),
    entity_id: str = Query(..., description="实体 ID"),
    status: str = Query(..., description="新状态"),
):
    """操作成功后同步更新 BIZ 库中的 status 字段"""
    if entity_type == "campaign":
        affected = await asyncio.to_thread(biz_campaign_repository.update_status, platform, entity_id, status)
    elif entity_type == "adgroup":
        affected = await asyncio.to_thread(biz_adgroup_repository.update_status, platform, entity_id, status)
    elif entity_type == "ad":
        affected = await asyncio.to_thread(biz_ad_repository.update_status, platform, entity_id, status)
    else:
        raise HTTPException(400, f"未知 entity_type: {entity_type}")
    return {"code": 0, "message": "ok", "data": {"affected": affected}}


@router.get("/adgroups")
async def biz_adgroups(
    platform: Optional[str] = Query(None, description="平台过滤: tiktok / meta"),
):
    """返回 BIZ 库中所有 adgroup（含 status / account_id）"""
    if platform:
        rows = await asyncio.to_thread(biz_adgroup_repository.list_by_platform, platform)
    else:
        rows = await asyncio.to_thread(biz_adgroup_repository.list_all)
    for row in rows:
        for k in ("created_at", "updated_at"):
            if k in row and row[k] is not None:
                row[k] = str(row[k])
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/ads")
async def biz_ads(
    platform: Optional[str] = Query(None, description="平台过滤: tiktok / meta"),
):
    """返回 BIZ 库中所有 ad（含 status / account_id）"""
    if platform:
        rows = await asyncio.to_thread(biz_ad_repository.list_by_platform, platform)
    else:
        rows = await asyncio.to_thread(biz_ad_repository.list_all)
    for row in rows:
        for k in ("created_at", "updated_at"):
            if k in row and row[k] is not None:
                row[k] = str(row[k])
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/overview")
async def biz_overview(
    start_date: str = Query(..., description="开始日期 YYYY-MM-DD"),
    end_date: str = Query(..., description="结束日期 YYYY-MM-DD"),
    platform: Optional[str] = Query(None, description="平台过滤: tiktok / meta"),
    account_id: Optional[str] = Query(None, description="账户 ID 过滤"),
):
    _check_dates(start_date, end_date)
    data = await asyncio.to_thread(
        biz_daily_report_repository.get_overview,
        start_date, end_date,
        platform=platform, account_id=account_id,
    )
    for k, v in data.items():
        if hasattr(v, "__float__"):
            data[k] = float(v)
        elif hasattr(v, "__int__") and not isinstance(v, bool):
            data[k] = int(v)
    return {"code": 0, "message": "ok", "data": data}


@router.get("/campaign-daily")
async def biz_campaign_daily(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    account_id: Optional[str] = Query(None),
    campaign_name: Optional[str] = Query(None, description="模糊搜索"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=200),
    order_by: str = Query("stat_date"),
    order_dir: str = Query("desc"),
):
    _check_dates(start_date, end_date)
    data = await asyncio.to_thread(
        biz_daily_report_repository.get_campaign_daily_list,
        start_date, end_date,
        platform=platform, account_id=account_id,
        campaign_name=campaign_name,
        page=page, page_size=page_size,
        order_by=order_by, order_dir=order_dir,
    )
    for row in data["list"]:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": data}


@router.get("/top-campaigns")
async def biz_top_campaigns(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    account_id: Optional[str] = Query(None),
    metric: str = Query("spend", description="排序指标: spend/revenue/clicks/installs/conversions/roas"),
    limit: int = Query(20, ge=1, le=100),
):
    _check_dates(start_date, end_date)
    rows = await asyncio.to_thread(
        biz_daily_report_repository.get_top_campaigns,
        start_date, end_date,
        platform=platform, account_id=account_id,
        metric=metric, limit=limit,
    )
    for row in rows:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/campaign-agg")
async def biz_campaign_aggregated(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    account_id: Optional[str] = Query(None),
    order_by: str = Query("total_spend"),
    order_dir: str = Query("desc"),
):
    """Campaign 聚合数据 — 按 campaign_id 分组，时间段内汇总"""
    _check_dates(start_date, end_date)
    rows = await asyncio.to_thread(
        biz_daily_report_repository.get_campaign_aggregated,
        start_date, end_date,
        platform=platform, account_id=account_id,
        order_by=order_by, order_dir=order_dir,
    )
    for row in rows:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/adgroup-agg")
async def biz_adgroup_aggregated(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    campaign_id: Optional[str] = Query(None, description="按 campaign 过滤"),
    order_by: str = Query("total_spend"),
    order_dir: str = Query("desc"),
):
    """Adgroup 聚合数据 — 按 adgroup_id 分组，支持按 campaign_id 过滤"""
    _check_dates(start_date, end_date)
    rows = await asyncio.to_thread(
        biz_adgroup_daily_repository.get_adgroup_aggregated,
        start_date, end_date,
        platform=platform, campaign_id=campaign_id,
        order_by=order_by, order_dir=order_dir,
    )
    for row in rows:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/ad-agg")
async def biz_ad_aggregated(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    campaign_id: Optional[str] = Query(None),
    adgroup_id: Optional[str] = Query(None, description="按 adgroup 过滤"),
    order_by: str = Query("total_spend"),
    order_dir: str = Query("desc"),
):
    """Ad 聚合数据 — 按 ad_id 分组，支持按 adgroup_id 过滤"""
    _check_dates(start_date, end_date)
    rows = await asyncio.to_thread(
        biz_ad_daily_repository.get_ad_aggregated,
        start_date, end_date,
        platform=platform, campaign_id=campaign_id, adgroup_id=adgroup_id,
        order_by=order_by, order_dir=order_dir,
    )
    for row in rows:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": rows}


@router.get("/adgroup-daily")
async def biz_adgroup_daily(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    account_id: Optional[str] = Query(None),
    name_filter: Optional[str] = Query(None, description="adgroup 名称模糊搜索"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=200),
    order_by: str = Query("stat_date"),
    order_dir: str = Query("desc"),
):
    _check_dates(start_date, end_date)
    data = await asyncio.to_thread(
        biz_adgroup_daily_repository.get_daily_list,
        start_date, end_date,
        platform=platform, account_id=account_id,
        name_filter=name_filter,
        page=page, page_size=page_size,
        order_by=order_by, order_dir=order_dir,
    )
    for row in data["list"]:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": data}


@router.get("/ad-daily")
async def biz_ad_daily_list(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    account_id: Optional[str] = Query(None),
    name_filter: Optional[str] = Query(None, description="ad 名称模糊搜索"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=200),
    order_by: str = Query("stat_date"),
    order_dir: str = Query("desc"),
):
    _check_dates(start_date, end_date)
    data = await asyncio.to_thread(
        biz_ad_daily_repository.get_daily_list,
        start_date, end_date,
        platform=platform, account_id=account_id,
        name_filter=name_filter,
        page=page, page_size=page_size,
        order_by=order_by, order_dir=order_dir,
    )
    for row in data["list"]:
        for k, v in row.items():
            if hasattr(v, "__float__") and not isinstance(v, (int, bool)):
                row[k] = float(v)
    return {"code": 0, "message": "ok", "data": data}


# ═══════════════════════ 素材分析 ═══════════════════════

@router.get("/creative-analysis")
async def creative_analysis(
    start_date: str = Query(...),
    end_date: str = Query(...),
    platform: Optional[str] = Query(None),
    min_spend: float = Query(10, description="低表现榜最小消耗门槛"),
    top_n: int = Query(5, ge=1, le=20),
):
    """素材分析综合接口 — 基于 ad 维度聚合，返回 overview / top / low / list"""
    _check_dates(start_date, end_date)

    agg_rows = await asyncio.to_thread(
        biz_ad_daily_repository.get_ad_aggregated,
        start_date, end_date,
        platform=platform,
        order_by="total_spend", order_dir="desc",
    )

    def _float(v):
        if v is None:
            return None
        return float(v)

    items = []
    for r in agg_rows:
        items.append({
            "ad_id": r.get("ad_id", ""),
            "ad_name": r.get("ad_name", ""),
            "platform": r.get("platform", ""),
            "impressions": int(r.get("total_impressions") or 0),
            "clicks": int(r.get("total_clicks") or 0),
            "spend": _float(r.get("total_spend")),
            "revenue": _float(r.get("total_revenue")),
            "installs": int(r.get("total_installs") or 0),
            "conversions": int(r.get("total_conversions") or 0),
            "ctr": _float(r.get("ctr")),
            "roas": _float(r.get("roas")),
        })

    total_count = len(items)
    total_spend = sum(i["spend"] or 0 for i in items)
    total_revenue = sum(i["revenue"] or 0 for i in items)
    total_impressions = sum(i["impressions"] for i in items)
    total_clicks = sum(i["clicks"] for i in items)
    avg_ctr = (total_clicks / total_impressions) if total_impressions > 0 else None
    avg_roas = (total_revenue / total_spend) if total_spend > 0 else None

    overview = {
        "total_creatives": total_count,
        "avg_ctr": round(avg_ctr, 6) if avg_ctr is not None else None,
        "avg_completion_rate": None,
        "avg_roas": round(avg_roas, 4) if avg_roas is not None else None,
        "total_spend": round(total_spend, 2),
        "total_revenue": round(total_revenue, 2),
    }

    top_by_roas = sorted(
        [i for i in items if (i["roas"] or 0) > 0],
        key=lambda x: x["roas"] or 0, reverse=True,
    )[:top_n]

    low_performers = sorted(
        [i for i in items if (i["spend"] or 0) >= min_spend],
        key=lambda x: x["roas"] if x["roas"] is not None else 999,
    )[:top_n]

    return {
        "code": 0,
        "message": "ok",
        "data": {
            "overview": overview,
            "top": top_by_roas,
            "low": low_performers,
            "list": items,
        },
    }

File: frontend/src/services/biz.ts
import { apiFetch } from './api'

// ---------------------------------------------------------------------------
// 公共参数类型
// ---------------------------------------------------------------------------

export interface BizDateRange {
  startDate: string            // YYYY-MM-DD
  endDate: string              // YYYY-MM-DD
  platform?: string            // tiktok | meta, 缺省 = 全部
}

// ---------------------------------------------------------------------------
// BIZ Campaigns (status / account_id from synced data)
// ---------------------------------------------------------------------------

export interface BizCampaign {
  id: number
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  objective: string
  buying_type: string
  status: string
  is_active: number
  raw_json: Record<string, unknown> | null
}

// ---------------------------------------------------------------------------
// Overview
// ---------------------------------------------------------------------------

export interface BizOverview {
  total_spend: number
  total_revenue: number
  total_impressions: number
  total_clicks: number
  total_installs: number
  total_conversions: number
  total_registrations?: number
  /**
   * Attribution 数仓侧自报的 spend 总和。仅 source=blend/auto 返回。
   * normalized 端 spend>0 但 attribution_spend≈0 → 数仓未采集该账号广告事件，
   * 此时 total_revenue / avg_roas 不可信，前端应给出"无归因数据"警示。
   */
  attribution_spend?: number
  avg_ctr: number | null
  avg_cpc: number | null
  avg_cpm: number | null
  avg_cpi: number | null
  avg_cpa: number | null
  avg_roas: number | null
}

// ---------------------------------------------------------------------------
// Top Campaigns
// ---------------------------------------------------------------------------

export interface BizTopCampaign {
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  total_spend: number
  total_impressions: number
  total_clicks: number
  total_installs: number
  total_conversions: number
  total_registrations?: number
  total_revenue: number
  avg_roas: number | null
}

export interface TopCampaignsParams extends BizDateRange {
  metric?: string              // spend | revenue | clicks | roas …
  limit?: number
}

// ---------------------------------------------------------------------------
// Campaign Daily
// ---------------------------------------------------------------------------

export interface BizCampaignDaily {
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  stat_date: string
  spend: number
  impressions: number
  clicks: number
  installs: number
  conversions: number
  registrations?: number
  revenue: number
  ctr: number | null
  cpc: number | null
  cpm: number | null
  cpi: number | null
  cpa: number | null
  roas: number | null
}

export interface CampaignDailyParams extends BizDateRange {
  page?: number
  page_size?: number
  order_by?: string
  order_dir?: 'asc' | 'desc'
}

// ---------------------------------------------------------------------------
// Adgroup Daily
// ---------------------------------------------------------------------------

export interface BizAdgroupDaily {
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  adgroup_id: string
  adgroup_name: string
  stat_date: string
  spend: number
  impressions: number
  clicks: number
  installs: number
  conversions: number
  revenue: number
  ctr: number | null
  cpc: number | null
  cpm: number | null
  cpi: number | null
  cpa: number | null
  roas: number | null
}

export interface AdgroupDailyParams extends BizDateRange {
  name_filter?: string
  page?: number
  page_size?: number
  order_by?: string
  order_dir?: 'asc' | 'desc'
}

// ---------------------------------------------------------------------------
// Ad Daily
// ---------------------------------------------------------------------------

export interface BizAdDaily {
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  adgroup_id: string
  adgroup_name: string
  ad_id: string
  ad_name: string
  stat_date: string
  spend: number
  impressions: number
  clicks: number
  installs: number
  conversions: number
  revenue: number
  ctr: number | null
  cpc: number | null
  cpm: number | null
  cpi: number | null
  cpa: number | null
  roas: number | null
}

export interface AdDailyParams extends BizDateRange {
  name_filter?: string
  page?: number
  page_size?: number
  order_by?: string
  order_dir?: 'asc' | 'desc'
}

export interface PaginatedResult<T> {
  total: number
  list: T[]
  page: number
  page_size: number
}

// ---------------------------------------------------------------------------
// Aggregated types (grouped, no per-day split)
// ---------------------------------------------------------------------------

export interface AggRow {
  platform: string
  account_id: string
  campaign_id: string
  campaign_name: string
  adgroup_id?: string
  adgroup_name?: string
  ad_id?: string
  ad_name?: string
  total_spend: number
  total_revenue: number
  total_impressions: number
  total_clicks: number
  total_installs: number
  total_conversions: number
  total_registrations?: number
  /** Attribution 数仓侧自报 spend；详见 BizOverview.attribution_spend 注释。 */
  attribution_spend?: number
  ctr: number | null
  cpc: number | null
  cpm: number | null
  cpi: number | null
  cpa: number | null
  roas: number | null
}

export interface AggParams extends BizDateRange {
  campaign_id?: string
  adgroup_id?: string
  order_by?: string
  order_dir?: 'asc' | 'desc'
}

// ---------------------------------------------------------------------------
// 内部工具
// ---------------------------------------------------------------------------

interface ApiResp<T> { code: number; message: string; data: T }

function qs(params: Record<string, string | number | undefined | null>): string {
  const parts: string[] = []
  for (const [k, v] of Object.entries(params)) {
    if (v != null && v !== '') parts.push(`${k}=${encodeURIComponent(v)}`)
  }
  return parts.length ? `?${parts.join('&')}` : ''
}

// ---------------------------------------------------------------------------
// API 函数（唯一出口）
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// BIZ Adgroups (status / account_id from synced data)
// ---------------------------------------------------------------------------

export interface BizAdgroup {
  id: number
  platform: string
  account_id: string
  campaign_id: string
  adgroup_id: string
  adgroup_name: string
  status: string
  is_active: number
  raw_json: Record<string, unknown> | null
}

// ---------------------------------------------------------------------------
// BIZ Ads (status / account_id from synced data)
// ---------------------------------------------------------------------------

export interface BizAd {
  id: number
  platform: string
  account_id: string
  campaign_id: string
  adgroup_id: string
  ad_id: string
  ad_name: string
  status: string
  is_active: number
  raw_json: Record<string, unknown> | null
}

export async function updateBizEntityStatus(
  platform: string,
  entityType: 'campaign' | 'adgroup' | 'ad',
  entityId: string,
  status: string,
): Promise<void> {
  await apiFetch<ApiResp<unknown>>(
    `/api/biz/update-status${qs({ platform, entity_type: entityType, entity_id: entityId, status })}`,
    { method: 'POST' },
  )
}

export async function fetchBizAdgroups(platform?: string): Promise<BizAdgroup[]> {
  const r = await apiFetch<ApiResp<BizAdgroup[]>>(
    `/api/biz/adgroups${qs({ platform })}`,
  )
  return r.data
}

export async function fetchBizAds(platform?: string): Promise<BizAd[]> {
  const r = await apiFetch<ApiResp<BizAd[]>>(
    `/api/biz/ads${qs({ platform })}`,
  )
  return r.data
}

export async function fetchBizCampaigns(platform?: string): Promise<BizCampaign[]> {
  const r = await apiFetch<ApiResp<BizCampaign[]>>(
    `/api/biz/campaigns${qs({ platform })}`,
  )
  return r.data
}

export async function fetchBizOverview(p: BizDateRange): Promise<BizOverview> {
  const r = await apiFetch<ApiResp<BizOverview>>(
    `/api/biz/overview${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
    })}`,
  )
  return r.data
}

export async function fetchBizTopCampaigns(p: TopCampaignsParams): Promise<BizTopCampaign[]> {
  const r = await apiFetch<ApiResp<BizTopCampaign[]>>(
    `/api/biz/top-campaigns${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      metric: p.metric,
      limit: p.limit,
    })}`,
  )
  return r.data
}

export async function fetchBizCampaignDaily(p: CampaignDailyParams): Promise<PaginatedResult<BizCampaignDaily>> {
  const r = await apiFetch<ApiResp<PaginatedResult<BizCampaignDaily>>>(
    `/api/biz/campaign-daily${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      page: p.page,
      page_size: p.page_size,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

export async function fetchBizAdgroupDaily(p: AdgroupDailyParams): Promise<PaginatedResult<BizAdgroupDaily>> {
  const r = await apiFetch<ApiResp<PaginatedResult<BizAdgroupDaily>>>(
    `/api/biz/adgroup-daily${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      name_filter: p.name_filter,
      page: p.page,
      page_size: p.page_size,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

export async function fetchBizAdDaily(p: AdDailyParams): Promise<PaginatedResult<BizAdDaily>> {
  const r = await apiFetch<ApiResp<PaginatedResult<BizAdDaily>>>(
    `/api/biz/ad-daily${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      name_filter: p.name_filter,
      page: p.page,
      page_size: p.page_size,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

// ---------------------------------------------------------------------------
// Aggregated fetches (for hierarchical expand)
// ---------------------------------------------------------------------------

export async function fetchCampaignAgg(p: AggParams): Promise<AggRow[]> {
  const r = await apiFetch<ApiResp<AggRow[]>>(
    `/api/biz/campaign-agg${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

export async function fetchAdgroupAgg(p: AggParams): Promise<AggRow[]> {
  const r = await apiFetch<ApiResp<AggRow[]>>(
    `/api/biz/adgroup-agg${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      campaign_id: p.campaign_id,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

export async function fetchAdAgg(p: AggParams): Promise<AggRow[]> {
  const r = await apiFetch<ApiResp<AggRow[]>>(
    `/api/biz/ad-agg${qs({
      start_date: p.startDate,
      end_date: p.endDate,
      platform: p.platform,
      campaign_id: p.campaign_id,
      adgroup_id: p.adgroup_id,
      order_by: p.order_by,
      order_dir: p.order_dir,
    })}`,
  )
  return r.data
}

// ── 素材分析 ──

export interface CreativeItem {
  ad_id: string
  ad_name: string
  platform: string
  impressions: number
  clicks: number
  spend: number | null
  revenue: number | null
  installs: number
  conversions: number
  ctr: number | null
  roas: number | null
  /** 剧维度：来自 ad_drama_mapping JOIN，未命中则为空串 */
  content_key?: string
  localized_drama_name?: string
  language_code?: string
  drama_id?: string
}

export interface CreativeOverview {
  total_creatives: number
  avg_ctr: number | null
  avg_completion_rate: number | null
  avg_roas: number | null
  total_spend: number
  total_revenue: number
}

export interface CreativeAnalysisData {
  overview: CreativeOverview
  top: CreativeItem[]
  low: CreativeItem[]
  list: CreativeItem[]
}

export interface CreativeAnalysisParams {
  startDate: string
  endDate: string
  platform?: string
  minSpend?: number
  topN?: number
  /** 剧筛选三件套（任意组合）*/
  contentKey?: string
  dramaKeyword?: string
  languageCode?: string
}

export async function fetchCreativeAnalysis(
  params: CreativeAnalysisParams,
): Promise<CreativeAnalysisData> {
  const r = await apiFetch<ApiResp<CreativeAnalysisData>>(
    `/api/biz/creative-analysis${qs({
      start_date: params.startDate,
      end_date: params.endDate,
      platform: params.platform,
      min_spend: params.minSpend,
      top_n: params.topN,
      content_key: params.contentKey,
      drama_keyword: params.dramaKeyword,
      language_code: params.languageCode,
    })}`,
  )
  return r.data
}

// ── 剧名筛选选项 ──

export interface DramaOption {
  content_key: string
  localized_drama_name: string
  /** 该剧在当前数据窗内涉及的所有语种（去重并按字母序） */
  language_codes: string[]
  total_spend: number
}

export interface DramaOptionsData {
  dramas: DramaOption[]
  /** 当前窗口内出现过投放的所有语种（用于"语种"下拉的全集） */
  languages: string[]
}

export async function fetchCreativeDramaOptions(params: {
  startDate: string
  endDate: string
  platform?: string
}): Promise<DramaOptionsData> {
  const r = await apiFetch<ApiResp<DramaOptionsData>>(
    `/api/biz/creative-analysis/drama-options${qs({
      start_date: params.startDate,
      end_date: params.endDate,
      platform: params.platform,
    })}`,
  )
  return r.data
}
